File: BackEnd/PHP/laravelshop/app/Http/Controllers/Api/V1/Matching/MatchingController.php
<?php

namespace App\Http\Controllers\Api\V1\Matching;

use OpenApi\Attributes as OA;

use App\Http\Controllers\Api\V1\ApiController;
use App\Modules\Matching\Requests\ApproveMappingRequest;
use App\Modules\Matching\Requests\RejectMappingRequest;
use App\Modules\Matching\Services\MatchingService;
use Illuminate\Http\Request;

#[OA\Tag(name: 'Matching')]
class MatchingController extends ApiController
{
    public function __construct(private readonly MatchingService $service) {}

    public function pending(Request $request)
    {
        return $this->success($this->service->pending($request->user(), (int) $request->query('page', 1), (int) $request->query('limit', 20)));
    }

    public function approve(ApproveMappingRequest $request, int $id)
    {
        return $this->success($this->service->approve($request->user(), $id, (int) $request->validated()['productId']));
    }

    public function reject(RejectMappingRequest $request, int $id)
    {
        return $this->success($this->service->reject($request->user(), $id, (string) $request->validated()['reason']));
    }

    public function autoMatch(Request $request)
    {
        return $this->success($this->service->autoMatch($request->user()));
    }

    public function stats(Request $request)
    {
        return $this->success($this->service->stats($request->user()));
    }
}

File: BackEnd/Kotlin/kotlin-ktor-gradle-exposeddao-postgresql/src/main/kotlin/com/pbshop/kotlin/ktor/gradle/exposeddao/postgresql/matching/MatchingController.kt
package com.pbshop.kotlin.ktor.gradle.exposeddao.postgresql.matching

import com.pbshop.kotlin.ktor.gradle.exposeddao.postgresql.common.pbActor
import com.pbshop.kotlin.ktor.gradle.exposeddao.postgresql.common.respondStub
import com.pbshop.kotlin.ktor.gradle.exposeddao.postgresql.security.PbRole
import com.pbshop.kotlin.ktor.gradle.exposeddao.postgresql.security.requireAnyRole
import io.ktor.server.application.call
import io.ktor.server.request.receive
import io.ktor.server.routing.Route
import io.ktor.server.routing.get
import io.ktor.server.routing.patch
import io.ktor.server.routing.post

class MatchingController(
    private val service: MatchingService,
) {
    fun Route.register() {
        get("/matching/pending") {
            call.requireAnyRole(PbRole.ADMIN)
            call.respondStub(
                service.pending(
                    page = call.request.queryParameters["page"]?.toIntOrNull() ?: 1,
                    limit = call.request.queryParameters["limit"]?.toIntOrNull() ?: 20,
                ),
            )
        }
        patch("/matching/{id}/approve") {
            call.requireAnyRole(PbRole.ADMIN)
            val request = call.receive<ApproveMappingRequest>()
            call.respondStub(service.approve(call.pathInt("id"), request.productId, call.currentUserId()))
        }
        patch("/matching/{id}/reject") {
            call.requireAnyRole(PbRole.ADMIN)
            val request = call.receive<RejectMappingRequest>()
            call.respondStub(service.reject(call.pathInt("id"), request.reason, call.currentUserId()))
        }
        post("/matching/auto-match") {
            call.requireAnyRole(PbRole.ADMIN)
            call.respondStub(service.autoMatch(call.currentUserId()))
        }
        get("/matching/stats") {
            call.requireAnyRole(PbRole.ADMIN)
            call.respondStub(service.stats())
        }
    }

    private fun io.ktor.server.application.ApplicationCall.currentUserId(): Int = pbActor().userId ?: 0

    private fun io.ktor.server.application.ApplicationCall.pathInt(name: String): Int = parameters[name]?.toIntOrNull() ?: 0
}
